### PR TITLE
Improve flutter_tools handling of the device being unplugged while syncing DevFS

### LINKED ISSUE #6705: `flutter run` doesn't handle the device going away during file syncing gracefully
Pull the cable from your phone a few seconds after the "Syncing files to device..." message appears:

```
$ flutter run
Installing build/app.apk...
Syncing files to device...
Application finished.
Error writing "build/flx/packages/flutter_gallery_assets/top_10_australian_beaches.jpg" to DevFS: HttpException: Connection closed before full header was received, uri = http://127.0.0.1:8100
dart:async/stream_controller.dart 590                      _StreamController.addError
dart:io/http_parser.dart 1070                              _HttpParser._reportError
dart:io/http_parser.dart 843                               _HttpParser._onDone
package:stack_trace/src/stack_zone_specification.dart 206  StackZoneSpecification._run
package:stack_trace/src/stack_zone_specification.dart 125  StackZoneSpecification.registerCallback.<fn>
package:stack_trace/src/stack_zone_specification.dart 206  StackZoneSpecification._run
package:stack_trace/src/stack_zone_specification.dart 125  StackZoneSpecification.registerCallback.<fn>
dart:async/zone.dart 1146                                  _rootRun
dart:async/zone.dart 1026                                  _CustomZone.run
dart:async/zone.dart 924                                   _CustomZone.runGuarded
dart:async/stream_impl.dart 384                            _BufferingStreamSubscription._sendDone.sendDone
dart:async/stream_impl.dart 394                            _BufferingStreamSubscription._sendDone
dart:async/stream_impl.dart 278                            _BufferingStreamSubscription._close
[...this goes on for over a million lines...]
```

### PR DESCRIPTION
**Pull request by johnmccutchan**
After this PR:

```
Launching lib/main.dart on Nexus 7 in debug mode...
Running 'gradle assembleDebug'...                     1.2s
Built build/app/outputs/apk/app-debug.apk (21.7MB).
Syncing files to device...
Application finished.
DevFS sync failed. Lost connection to device: SocketException: OS Error: Connection refused, errno = 111, address = 127.0.0.1, port = 53062
Could not perform initial file synchronization.
```

Fixes #6705

### CODE REVIEW

File: packages/flutter_tools/lib/src/devfs.dart [hunk 1]
```diff
@@ -274,11 +281,17 @@ class _DevFSHttpWriter {
...
       request.headers.add('dev_fs_uri_b64',
-                          BASE64.encode(UTF8.encode(deviceUri.toString())));
+          BASE64.encode(UTF8.encode(deviceUri.toString())));
       final Stream<List<int>> contents = content.contentsAsCompressedStream();
       await request.addStream(contents);
       final HttpClientResponse response = await request.close();
       await response.drain<Null>();
+    } on SocketException catch (socketException, stackTrace) {
+      // We have one completer and can get up to kMaxInFlight errors.
+      if (!_completer.isCompleted) {
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/devfs.dart]: nit: braces

**Comment by Hixie** [comment]
The code LGTM.
Maybe a test?

I would make a couple of tweaks to the output. The first is that I wouldn't mention DevFS. Nobody outside our team knows what that is. The other is that if we are pretty sure the error is that we lost the device connection, there's no need to output any of the exception data, IP addresses, ports, and so on. We can dump those to --verbose if they're helpful, but in regular output I would just keep it simple. Similarly, if we can drop the "Application finished" in the lost-connection case that'd be ideal -- after all, we don't know that it finished -- and there's probably no point saying "Could not perform initial file synchronization" since we're about to give up entirely and quit, meaning that the user has no expectation of us doing anything, let alone any initial file sync or anything.

That is, I would aim for this output for this case:
```
$ flutter run
Launching lib/main.dart on Nexus 7 in debug mode...
Running 'gradle assembleDebug'...                     1.2s
Built build/app/outputs/apk/app-debug.apk (21.7MB).
Syncing files to device...                         failed.
Lost connection to device.
$
```

**Comment by Hixie** [comment]
(but those changes are long-term; feel free to just file an issue to look at this more later. the change here is a clear improvement and LGTM as is.)

**Comment by johnmccutchan** [comment]
I've made some tweaks based on your feedback.

```
Launching lib/main.dart on Nexus 7 in debug mode...
Running 'gradle assembleDebug'...                     1.2s
Built build/app/outputs/apk/app-debug.apk (21.7MB).
Syncing files to device...                                
Lost connection to device.
```

**Pull request merged**